File: src/main/java/com/now/nowbot/service/MessageServiceImpl/MapMinusService.java
package com.now.nowbot.service.MessageServiceImpl;

import com.now.nowbot.model.JsonData.BeatMap;
import com.now.nowbot.model.beatmapParse.OsuFile;
import com.now.nowbot.model.enums.OsuMod;
import com.now.nowbot.model.enums.OsuMode;
import com.now.nowbot.model.ppminus3.PPMinus3;
import com.now.nowbot.qq.event.MessageEvent;
import com.now.nowbot.service.ImageService;
import com.now.nowbot.service.MessageService;
import com.now.nowbot.service.OsuApiService.OsuBeatmapApiService;
import com.now.nowbot.throwable.ServiceException.MapMinusException;
import com.now.nowbot.util.DataUtil;
import com.now.nowbot.util.Instruction;
import jakarta.annotation.Resource;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;

import java.util.regex.Matcher;

import static com.now.nowbot.util.command.CommandPatternStaticKt.FLAG_MOD;

@Service("MAP_MINUS")
public class MapMinusService implements MessageService<Matcher> {
    private static final Logger log = LoggerFactory.getLogger(MapMinusService.class);
    @Resource
    OsuBeatmapApiService beatmapApiService;
    @Resource
    ImageService imageService;

    @Override
    public boolean isHandle(MessageEvent event, String messageText, DataValue<Matcher> data) {
        var m = Instruction.MAP_MINUS.matcher(messageText);
        if (m.find()) {
            data.setValue(m);
            return true;
        } else return false;
    }

    @Override
    public void HandleMessage(MessageEvent event, Matcher matcher) throws Throwable {
        var from = event.getSubject();
        long bid;
        double rate = 1d;
        OsuMode mode;
        String fileStr;
        BeatMap beatMap;

        var modsList = OsuMod.getModsList(matcher.group(FLAG_MOD));
        boolean isChangedRating = OsuMod.hasChangeRating(modsList);


        try {
            bid = Long.parseLong(matcher.group("bid"));
        } catch (NumberFormatException e) {
            throw new MapMinusException(MapMinusException.Type.MM_Bid_Error);
        }

        if (StringUtils.hasText(matcher.group("rate"))) {
            try {
                rate = Double.parseDouble(matcher.group("rate"));
            } catch (NumberFormatException e) {
                throw new MapMinusException(MapMinusException.Type.MM_Rate_Error);
            }
        }

        if (rate < 0.1d) throw new MapMinusException(MapMinusException.Type.MM_Rate_TooSmall);
        if (rate > 5d) throw new MapMinusException(MapMinusException.Type.MM_Rate_TooLarge);


        try {
            beatMap = beatmapApiService.getBeatMapInfoFromDataBase(bid);
            mode = OsuMode.getMode(beatMap.getModeInt());
            int mods = OsuMod.getModsValue(modsList);
            var s = beatmapApiService.getMaxPP(beatMap.getBeatMapID(), mods);
            beatMap.setStarRating((float) s.getStar());
            DataUtil.applyBeatMapChanges(beatMap, mods);
            fileStr = beatmapApiService.getBeatMapFile(bid);
        } catch (Exception e) {
            throw new MapMinusException(MapMinusException.Type.MM_Map_NotFound);
        }

        OsuFile file = null;
        if (mode != null) {
            try {
                switch (mode) {
                    case MANIA -> file = OsuFile.getInstance(fileStr);
                    default -> throw new MapMinusException(MapMinusException.Type.MM_Function_NotSupported);
                    //throw new TipsException("抱歉，本功能暂不支持除Mania模式以外的谱面！");//file = new OsuFile(fileStr);
                }

            } catch (NullPointerException e) {
                throw new MapMinusException(MapMinusException.Type.MM_Map_FetchFailed);
            }
        }

        PPMinus3 mapMinus = null;
        if (file != null) {
            mapMinus = PPMinus3.getInstance(file, isChangedRating ? OsuMod.getModsClockRate(modsList) : rate);
        }

        byte[] image;

        try {
            image = imageService.getPanelB2(beatMap, mapMinus);
        } catch (Exception e) {
            log.error("谱面 Minus：渲染失败", e);
            throw new MapMinusException(MapMinusException.Type.MM_Render_Error);
        }

        try {
            from.sendImage(image);
        } catch (Exception e) {
            log.error("谱面 Minus：发送失败", e);
            throw new MapMinusException(MapMinusException.Type.MM_Send_Error);
        }
    }
}

File: src/main/java/com/now/nowbot/service/messageServiceImpl/MapMinusService.kt
package com.now.nowbot.service.messageServiceImpl

import com.now.nowbot.dao.ServiceCallStatisticsDao
import com.now.nowbot.dao.SkillDao
import com.now.nowbot.entity.ServiceCallStatistic
import com.now.nowbot.model.osu.LazerMod
import com.now.nowbot.model.beatmapParse.OsuFile
import com.now.nowbot.model.enums.OsuMode
import com.now.nowbot.model.osu.Beatmap
import com.now.nowbot.model.osu.LazerMod.Companion.isAffectStarRating
import com.now.nowbot.model.skill.Skill6
import com.now.nowbot.qq.event.MessageEvent
import com.now.nowbot.qq.message.MessageChain
import com.now.nowbot.qq.message.MessageChain.MessageChainBuilder
import com.now.nowbot.qq.tencent.TencentMessageService
import com.now.nowbot.service.ImageService
import com.now.nowbot.service.MessageService
import com.now.nowbot.service.MessageService.DataValue
import com.now.nowbot.service.osuApiService.OsuBeatmapApiService
import com.now.nowbot.service.osuApiService.OsuCalculateApiService
import com.now.nowbot.throwable.botRuntimeException.IllegalArgumentException
import com.now.nowbot.throwable.botRuntimeException.IllegalStateException
import com.now.nowbot.throwable.botRuntimeException.NoSuchElementException
import com.now.nowbot.throwable.botRuntimeException.UnsupportedOperationException
import com.now.nowbot.util.Instruction
import com.now.nowbot.util.OfficialInstruction
import com.now.nowbot.util.command.FLAG_MOD
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.time.LocalDateTime
import java.util.regex.Matcher
import kotlin.math.absoluteValue

@Service("MAP_MINUS") class MapMinusService(
    private val beatmapApiService: OsuBeatmapApiService,
    private val calculateApiService: OsuCalculateApiService,
    private val imageService: ImageService,
    private val skillDao: SkillDao,
    private val dao: ServiceCallStatisticsDao,
) : MessageService<MapMinusService.MapMinusParam>, TencentMessageService<MapMinusService.MapMinusParam> {

    data class MapMinusParam(val bid: Long, val mode: OsuMode, val rate: Double = 1.0, val mods: List<LazerMod>)

    override fun isHandle(
        event: MessageEvent,
        messageText: String,
        data: DataValue<MapMinusParam>,
    ): Boolean {
        val m = Instruction.MAP_MINUS.matcher(messageText)
        if (!m.find()) {
            return false
        }

        data.value = getMapMinusParam(event, m)
        return true
    }

    @Throws(Throwable::class) override fun handleMessage(event: MessageEvent, param: MapMinusParam): ServiceCallStatistic? {
        val image = getMapMinusImage(param, beatmapApiService, calculateApiService, imageService)

        try {
            event.reply(image)
        } catch (e: Exception) {
            log.error("谱面 Minus：发送失败", e)
            throw IllegalStateException.Send("谱面 Minus")
        }

        return ServiceCallStatistic.build(event, beatmapID = param.bid, mode = param.mode)
    }

    override fun accept(event: MessageEvent, messageText: String): MapMinusParam? {
        val m = OfficialInstruction.MAP_MINUS.matcher(messageText)
        if (!m.find()) {
            return null
        }

        return getMapMinusParam(event, m)
    }

    override fun reply(event: MessageEvent, param: MapMinusParam): MessageChain? {
        val image = getMapMinusImage(param, beatmapApiService, calculateApiService, imageService)
        return MessageChainBuilder().addImage(image).build()
    }

    private fun getMapMinusParam(event: MessageEvent, matcher: Matcher): MapMinusParam {
        val modsList: List<LazerMod> = LazerMod.getModsList(matcher.group(FLAG_MOD))

        val bid = matcher.group("bid")?.toLongOrNull()
            ?: dao.getLastBeatmapID(event.subject.contactID, name = null, LocalDateTime.now().minusHours(24))
            ?: throw IllegalArgumentException.WrongException.BeatmapID()

        val rate = matcher.group("rate")?.toDoubleOrNull() ?: 1.0

        if (rate < 0.1) throw IllegalArgumentException.ExceedException.RateTooSmall()
        if (rate > 5.0) throw IllegalArgumentException.ExceedException.RateTooLarge()

        return MapMinusParam(bid, OsuMode.MANIA, rate, modsList)
    }

    private fun getMapMinusImage(
        param: MapMinusParam,
        beatmapApiService: OsuBeatmapApiService,
        calculateApiService: OsuCalculateApiService,
        imageService: ImageService,
    ): ByteArray {
        val fileStr: String
        val map: Beatmap = beatmapApiService.getBeatmap(param.bid)

        calculateApiService.applyStarToBeatmap(map, param.mode, param.mods)

        val isChangedRating = param.mods.isAffectStarRating()

        try {
            fileStr = beatmapApiService.getBeatmapFileString(param.bid)!!
        } catch (_: Exception) {
            throw NoSuchElementException.BeatmapCache(param.bid)
        }

        if (map.mode.isNotConvertAble(param.mode)) {
            throw UnsupportedOperationException.OnlyMania()
        }

        val file = try {
            OsuFile(fileStr)
        } catch (_: NullPointerException) {
            throw IllegalStateException.Fetch("谱面文件")
        }

        val clockRate = if (isChangedRating) {
            LazerMod.getModSpeedForStarCalculate(param.mods).toDouble()
        } else {
            param.rate
        }

        val mapMinus = Skill6(
            file,
            param.mode,
            clockRate,
        )

        // val type = SkillType.getType(mapMinus)

        if ((clockRate - 1.0).absoluteValue <= 1e-4) {
            // skillDao.saveAndUpdateSkill(map, param.mode, mapMinus)
        }

        val body = mapOf(
            "beatmap" to map,
            "map_minus" to mapMinus,
            "type" to "null",//type.keys.first().name,
            "type_percent" to 0//type.values.first()
        )

        return imageService.getPanel(body, "B2")
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(MapMinusService::class.java)
    }
}
